packages/mdc-textfield/index.js

```diff
@@ -120,6 +109,23 @@ class MDCTextField extends MDCComponent {
     if (iconElement) {
       this.icon_ = iconFactory(iconElement);
     }
+
+    this.ripple = null;
+    if (this.root_.classList.contains(cssClasses.BOX) || this.root_.classList.contains(cssClasses.OUTLINED)) {
+      const MATCHES = getMatchesProperty(HTMLElement.prototype);
+      const rippleInputMethods = {
+        isSurfaceActive: () => this.input_[MATCHES](':active'),
+        registerInteractionHandler: (type, handler) => this.input_.addEventListener(type, handler),
+        deregisterInteractionHandler: (type, handler) => this.input_.removeEventListener(type, handler),
+      };
+      if (this.root_.classList.contains(cssClasses.BOX)) {
+        const adapter = Object.assign(MDCRipple.createAdapter(this), rippleInputMethods);
+        const foundation = new MDCRippleFoundation(adapter);
+        this.ripple = rippleFactory(this.root_, foundation);
+      } else if (this.root_.classList.contains(cssClasses.OUTLINED)) {
+        this.ripple = this.outline_.createRipple(rippleFactory, rippleInputMethods);
```

Comment:
It seems a bit overboard to me to have a separate method on the outline subcomponent that does exactly the same thing as above, and requiring us to pass everything through to it, when the only difference is which element the ripple is instantiated on.

I also find it a bit strange since the ripple is still destroyed from the top level component, and there's nothing really "owning" it from the outline subcomponent.

Given that the branching logic is here anyway, should we simply conditionalize which element the ripple is created on here and otherwise have the same code for both in one place?

Another thought: should the decision of which element hosts the ripple be determined within an API for the top-level foundation, so that component implementors have less logic to worry about?

Reply:
The reason for the separate method on the outline subcomponent is because we need access to the outline's root element to instantiate the ripple on. The other option was to have a public `getRoot()` method in the outline, which seems to break best practices....although that might seem less weird than having an entire `createRipple()` method on a subcomponent that doesn't even own the ripple. Thoughts?

Moving the ripple host decision to the foundation seems even more complicated since the main foundation would have to somehow access the outline subcomponent (don't know if this is possible, maybe through the outline subadapter?...)

Reply:
I would argue that we already have access to the outline subcomponent's root element: `outlineElement`. We just passed it _to_ the subcomponent constructor about 30 lines above in this same method.

And...yeah my idea about making the foundation decide ends up being pretty awkward, so nevermind about that one.

Reply:
D'oh, huge brain fart on my part. Updated it to share the same code except for conditionalizing on which element the ripple is created on.